File: backend/app/calc/weight.py
import math
from typing import List, Dict, Any

DIVISOR = {
    'AIR': 6000,
    'EXPRESS_AIR': 5000,
    'GROUND_RAIL': 4500,
    'OCEAN_LCL': 1000  # cm3 to CBM for revenue tons
}

CONTAINER_PAYLOAD_LIMITS = {
    '20GP': 21800,
    '40GP': 26600,
    '40HC': 28800,
    '20RF': 21000,
    '40RF': 27000,
    '20OT': 21500,
    '40FR': 31000
}

def actual_weight(items: List[Dict[str, Any]]) -> float:
    """Calculates total physical actual weight in kg."""
    tot = 0.0
    for item in items or []:
        if item.get('package_type') == 'CONTAINER':
            tot += float(item.get('gross_weight_kg', 0))
        else:
            qty = float(item.get('quantity', item.get('qty', 1)))
            w = float(item.get('weight_per_unit_kg', item.get('weight', 0)))
            tot += qty * w
    return round(tot, 2)

def volumetric_weight(items: List[Dict[str, Any]], mode: str = 'AIR') -> float:
    """Calculates total dimensional weight in kg."""
    mode_upper = mode.upper()
    div = DIVISOR.get(mode_upper, DIVISOR['AIR'])
    tot = 0.0
    for item in items or []:
        if item.get('package_type') == 'CONTAINER':
            continue
        qty = float(item.get('quantity', item.get('qty', 1)))
        l = float(item.get('length_cm', item.get('length', 0)))
        w = float(item.get('width_cm', item.get('width', 0)))
        h = float(item.get('height_cm', item.get('height', 0)))
        if l > 0 and w > 0 and h > 0:
            tot += (l * w * h * qty) / div
    return round(tot, 2)

def chargeable_weight(items: List[Dict[str, Any]], mode: str = 'OCEAN', load_type: str = 'FCL') -> Dict[str, Any]:
    """
    Computes Chargeable Weight & Basis across 3 branches:
    1. PER_CONTAINER (FCL)
    2. REVENUE_TON (LCL)
    3. CHARGEABLE_KG (Air, Express, Ground)
    """
    safe_items = items or []
    is_container = load_type.upper() == 'FCL' or any(i.get('package_type') == 'CONTAINER' for i in safe_items)

    if is_container:
        count = sum(int(i.get('container_count', 1)) for i in safe_items) or 1
        container_types = list(set(i.get('container_type', '40HC') for i in safe_items))
        types_str = ", ".join(container_types) if container_types else "40HC"
        return {
            'basis': 'PER_CONTAINER',
            'unitsLabel': f"{count} × {types_str}",
            'units': count,
            'chargeableVal': count
        }
    elif mode.upper() == 'OCEAN' and load_type.upper() == 'LCL':
        act_kg = actual_weight(safe_items)
        tonnes = act_kg / 1000.0
        cbm = sum((float(i.get('length_cm', 0)) * float(i.get('width_cm', 0)) * float(i.get('height_cm', 0)) * float(i.get('quantity', 1))) / 1000000.0 for i in safe_items)
        revenue_tons = max(cbm, tonnes, 1.0)
        return {
            'basis': 'REVENUE_TON',
            'unitsLabel': f"{revenue_tons:.1f} R/T",
            'units': round(revenue_tons, 2),
            'chargeableVal': round(revenue_tons, 2)
        }
    else:
        act_kg = actual_weight(safe_items)
        vol_kg = volumetric_weight(safe_items, mode)
        chg_kg = max(act_kg, vol_kg, 1.0)
        return {
            'basis': 'CHARGEABLE_KG',
            'unitsLabel': f"{math.ceil(chg_kg)} kg ch.",
            'units': math.ceil(chg_kg),
            'chargeableVal': math.ceil(chg_kg)
        }

File: frontend/src/lib/pricing/weight.js
import { DIVISORS, ContainerPayloadLimits } from './constants'

/**
 * Calculates total physical gross weight in kg
 */
export function actualWeight(items = []) {
  if (!Array.isArray(items)) return 0
  return items.reduce((sum, item) => {
    if (!item) return sum
    if (item.package_type === 'CONTAINER') {
      const weight = parseFloat(item.gross_weight_kg) || 0
      return sum + weight
    } else {
      const qty = parseFloat(item.quantity || item.qty) || 1
      const wPerUnit = parseFloat(item.weight_per_unit_kg || item.weight) || 0
      const gross = parseFloat(item.gross_weight_kg) || 0
      const unitWeight = wPerUnit > 0 ? wPerUnit : (gross > 0 ? gross / qty : 10)
      return sum + (qty * unitWeight)
    }
  }, 0)
}

/**
 * Calculates total volumetric weight in kg
 */
export function volumetricWeight(items = [], mode = 'AIR') {
  if (!Array.isArray(items)) return 0
  const divisor = DIVISORS[mode] || DIVISORS.AIR
  return items.reduce((sum, item) => {
    if (!item || item.package_type === 'CONTAINER') return sum
    
    const qty = parseFloat(item.quantity || item.qty) || 1
    const l = parseFloat(item.length_cm || item.length) || 0
    const w = parseFloat(item.width_cm || item.width) || 0
    const h = parseFloat(item.height_cm || item.height) || 0
    
    if (l <= 0 || w <= 0 || h <= 0) return sum
    const volCm3 = l * w * h * qty
    return sum + (volCm3 / divisor)
  }, 0)
}

/**
 * Computes Chargeable Weight & Basis
 * @returns {Object} { basis: string, isContainer: boolean, unitsLabel: string, units: number, chargeableVal: number }
 */
export function chargeableWeight(items = [], mode = 'OCEAN', loadType = 'FCL') {
  const safeItems = Array.isArray(items) ? items : []
  
  // Package is container ONLY if package_type is explicitly CONTAINER or (in Ocean mode with loadType FCL and non-specific package type)
  const hasContainerPackage = safeItems.some(i => i?.package_type === 'CONTAINER')
  const hasNonContainerPackage = safeItems.some(i => i?.package_type && i.package_type !== 'CONTAINER')
  
  const isContainer = mode === 'OCEAN' && (hasContainerPackage || (loadType === 'FCL' && !hasNonContainerPackage))

  if (isContainer) {
    let containerCount = safeItems.reduce((sum, i) => {
      const count = parseInt(i?.container_count) || 0
      if (count > 0) return sum + count
      const gross = parseFloat(i?.gross_weight_kg) || 0
      const limit = ContainerPayloadLimits[i?.container_type || '40HC'] || 28800
      return sum + Math.max(1, Math.ceil(gross / limit))
    }, 0)

    if (containerCount < 1) containerCount = 1

    const containerTypes = Array.from(new Set(safeItems.map(i => i?.container_type || '40HC'))).join(', ')
    return {
      basis: 'PER_CONTAINER',
      isContainer: true,
      unitsLabel: `${containerCount} × ${containerTypes || '40HC'}`,
      units: containerCount,
      chargeableVal: containerCount
    }
  } else if (mode === 'OCEAN' && (loadType === 'LCL' || hasNonContainerPackage)) {
    // Ocean LCL: charged on Revenue Tons: max(CBM, Metric Tonnes)
    const totActualKg = actualWeight(safeItems)
    const tonnes = totActualKg / 1000
    
    const cbm = safeItems.reduce((sum, i) => {
      const qty = parseFloat(i?.quantity || i?.qty) || 1
      const l = parseFloat(i?.length_cm || i?.length) || 0
      const w = parseFloat(i?.width_cm || i?.width) || 0
      const h = parseFloat(i?.height_cm || i?.height) || 0
      if (l <= 0 || w <= 0 || h <= 0) {
        // Fallback default CBM estimation based on package type if dimension not specified
        const defCbm = i?.package_type === 'PALLET' ? 1.2 : 0.2
        return sum + (defCbm * qty)
      }
      return sum + ((l * w * h * qty) / 1000000)
    }, 0)

    const revenueTons = Math.max(cbm, tonnes, 0.5)
    return {
      basis: 'REVENUE_TON',
      isContainer: false,
      unitsLabel: `${revenueTons.toFixed(1)} R/T`,
      units: Number(revenueTons.toFixed(2)),
      chargeableVal: Number(revenueTons.toFixed(2))
    }
  } else {
    // Air, Express Air, Ground & Rail: CHARGEABLE_KG = max(actual_weight, volumetric_weight)
    const actKg = actualWeight(safeItems)
    const volKg = volumetricWeight(safeItems, mode)
    const chgKg = Math.max(actKg, volKg, 1.0)

    return {
      basis: 'CHARGEABLE_KG',
      isContainer: false,
      unitsLabel: `${Math.ceil(chgKg)} kg ch.`,
      units: Math.ceil(chgKg),
      chargeableVal: Math.ceil(chgKg)
    }
  }
}
